examples/IBLevelSet/ex0/GravityForcing.cpp

```diff
@@ -0,0 +1,125 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "GravityForcing.h"
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <IBAMR_config.h>
+#include <SAMRAI_config.h>
+
+// SAMRAI INCLUDES
+#include <HierarchyDataOpsManager.h>
+
+/////////////////////////////// STATIC ///////////////////////////////////////
+
+/////////////////////////////// PUBLIC ///////////////////////////////////////
+
+GravityForcing::GravityForcing(const std::string& object_name,
+                               Pointer<INSVCStaggeredHierarchyIntegrator> ins_hierarchy_integrator,
+                               std::vector<double> grav_const)
+    : d_object_name(object_name), d_ins_hierarchy_integrator(ins_hierarchy_integrator), d_grav_const(grav_const)
+{
+    // intentionally blank
+    return;
+} // GravityForcing
+
+GravityForcing::~GravityForcing()
+{
+    // intentionally blank
+    return;
+} // ~GravityForcing
+
+bool
+GravityForcing::isTimeDependent() const
+{
+    return true;
+} // isTimeDependent
+
+void
+GravityForcing::setDataOnPatchHierarchy(const int data_idx,
+                                        Pointer<hier::Variable<NDIM> > /*var*/,
+                                        Pointer<PatchHierarchy<NDIM> > hierarchy,
+                                        const double /*data_time*/,
+                                        const bool /*initial_time*/,
+                                        const int coarsest_ln_in,
+                                        const int finest_ln_in)
+{
+    const int coarsest_ln = (coarsest_ln_in == -1 ? 0 : coarsest_ln_in);
+    const int finest_ln = (finest_ln_in == -1 ? hierarchy->getFinestLevelNumber() : finest_ln_in);
+
+    // Get interpolated density variable
+    const int rho_ins_idx = d_ins_hierarchy_integrator->getLinearOperatorRhoPatchDataIndex();
+
+#if !defined(NDEBUG)
+    TBOX_ASSERT(rho_ins_idx >= 0);
+#endif
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            Pointer<Patch<NDIM> > patch = level->getPatch(p());
```

Comment:
@amneetb, does this need to be here, or could it go in `setDataOnPatch()`?

Reply:
Since we need to fill ghost cells of level set, we do this on the entire hierarchy directly.

Reply:
I think we should make it possible to have the `CartGridFunction` base class fill ghost cells (in another PR).